### DISCUSSION TITLE
Waveshare SPI 2.4" LCD

### DISCUSSION THREAD

**Discussion by Westcott1** [Bodmer/TFT_eSPI#952] [Q&A - General]
Hi,
I have a Waveshare 2.4" LCD (ILI9341) that I'm trying to run from an ESP32 Devkit V1 (30 pins)
https://www.waveshare.com/wiki/2.4inch_LCD_Module
It's a blue-ish PCB with unusual pinout -
DIN (MOSI), CLK (SCLK), CS, DC, RST, BL
There is NO MISO pin, and according to Waveshare -
 "It is not like the traditional SPI protocol, it only uses MOSI to send data from master to slave for the LCD display"

Do you have any idea how to connect and control this display?
I'm going to try -

#define TFT_SDA_READ 
DIN  13
CLK  14
CS    15
DC   26
RST  27
BL    32

[code]
TFT_eSPI ver = 2.3.59
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9341
Display width  = 240
Display height = 320

MOSI    = GPIO 13
MISO    = GPIO 19
SCK     = GPIO 14
TFT_DC   = GPIO 26
TFT_RST  = GPIO 27

TFT_BL           = GPIO 32
TFT_BACKLIGHT_ON = HIGH

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 40.00
[/code]

Backlight control works for ON/OFF and PWM.
Screen remains obstinately white.

Thanks.

**Comment by Bodmer** [comment]
You do not need MISO so that pin at the ESP32 can be left unconnected.

This is the[ pinout and setup file](https://github.com/Bodmer/TFT_eSPI/discussions/950#discussioncomment-305458) I suggest you use. Those pins are the "native" SPI pins and are more reliable at high SPI clock rates.

**Comment by Bodmer** [comment]
You can either tie the BL pin to 3.3V to start with or define the TFT_BL pin as you suggest. I usualy try the simplest possible setup first of only display and ESP32 to eliminate any ardware conflicts that can arise.

**Comment by Westcott1** [comment]
Bodmer, thanks!
I had already left MISO unconnected, and backlight control worked correctly and dimming with PWM.
Unfortunately, swapping to your suggested pins hasn't helped, the screen is still just white (but PWM dimmed)
Could it be anything to do with RESX, CSX, or D/CX?

> RESX: Reset, should be pull-down when power on, set to 1 other time.
CSX: Slave chip select. The chip is enabled only CS is set Low
D/CX: Data/Command selection; DC=0, write command; DC=1, write data

**Comment by Bodmer** [comment]
The library manages those pins.

**Reply by Westcott1** [reply]
Thanks,
Waveshare provide an Arduino Uno project for this specific display.
I'll try and follow how it uses its pins.

**Reply by Bodmer** [reply]
Connecting to an UNO would not be a good idea because the display is 3.3V logic only and requires a 3.3V power supply.  You may damage it if you connect the UNO 5V logic signals to it even if the display is correctly powered from 3.3V. Unfortunately Chinese vendors do not tell you that logic level shifters are required with an UNO! [See here](https://www.instructables.com/Arduino-TFT-display-and-font-library/).

**Comment by Westcott1** [comment]
Hi again,
No luck using your pins, still a white screen.
I eventually got the Waveshare example project to compile and run on the ESP32, again just a white screen.
Also got the Arduino Adafruit GFX example to run - white screen.
I also ran your register read example -

TFT driver register values:
-===========================
Register 0x01: 0x00
Register 0x04: 0xFFFFFF
Register 0x09: 0xFFFFFFFF
Register 0x0A: 0xFF
Register 0x0B: 0xFF
Register 0x0C: 0xFF
Register 0x0D: 0xFF
Register 0x0E: 0xFF
Register 0x0F: 0xFF
Register 0x2E: 0xFFFFFF
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0xFF
Register 0xDB: 0xFF
Register 0xDC: 0xFF
-===========================

Thanks.

**Comment by Bodmer** [comment]
Did you run the "Read_ID_Bitbash" example that is in the library example set? I ask because I posted a similar sketch on another thread and that other one definitely would not work as it requires a MISO pin, which your display does not have.

**Reply by Westcott1** [reply]
I'll have a look for it.
Thanks for all of your help.

**Reply by Westcott1** [reply]
Ran the example, and added the value of ID -

TFT driver register values:
-===========================
Register 0x01: 0x00
Register 0x04: 0x00
Register 0x09: 0x610000
Register 0x0A: 0x08
Register 0x0B: 0x00
Register 0x0C: 0x06
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x545454
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0x00
Register 0xDB: 0x00
Register 0xDC: 0x00
-===========================

Looks like driver chip is: Unknown                                 
ID: 0

**Reply by Bodmer** [reply]
That result looks good. I would now try the TFT_eSPI library with the SPI clock rate set to 10MHz , I.e. 10000000

**Comment by Bodmer** [comment]
Note that to use ADafruit_GFX you must define all the pins (except MISO in your case):

```
Adafruit_ILI9341 tft = Adafruit_ILI9341(TFT_CS, TFT_DC, TFT_MOSI, TFT_CLK, TFT_RST, TFT_MISO);

```

**Reply by Westcott1** [reply]
#define _cs   15   // goes to TFT CS
#define _dc   16   // goes to TFT DC
#define _mosi 23  // goes to TFT MOSI
#define _sclk 18  // goes to TFT SCK/CLK
#define _rst  4  // ESP RST to TFT RESET
#define _miso     // Not connected
//       3.3V     // Goes to TFT LED  
//       5v       // Goes to TFT Vcc
//       Gnd      // Goes to TFT Gnd        

**Adafruit_ILI9341 tft = Adafruit_ILI9341(_cs, _dc, _mosi, _sclk, _rst);**

**Reply by Bodmer** [reply]
That would work but screen updates will be painfully slow as the Adafruit library will bit-bash the SPI bus.  The library assumes that since the MOSI and SCK signals are defined then the faster hardware SPI is not wanted. For some processors this is a reasonable approach, but the ESP32 SPI ports can be routed to different pins.

Use:
```
Adafruit_ILI9341 tft = Adafruit_ILI9341(_cs, _dc, _rst);
```
You can then control the SPI clock rate is you wish with for example:
```
  tft.begin(20000000); // 20MHz SPI clock
```

**Reply by Westcott1** [reply]
Wow, that really speeded-up the draw rates!
Unfortunately the screen remains white.

**Reply by Bodmer** [reply]
The Adaruit graphics test outputs 5 register values at the start, these show if your SPI bus is connected correctly. Here is my output with a SPI ILI9341 connected at 20MHz SPI clock rate:
```
Display Power Mode: 0x94
MADCTL Mode: 0x48
Pixel Format: 0x5
Image Format: 0x80
Self Diagnostic: 0xC0
Benchmark                Time (microseconds)
Screen fill              349306
Text                     35764
Lines                    345818
Horiz/Vert Lines         30154
Rectangles (outline)     19679
Rectangles (filled)      725254
Circles (filled)         113720
Circles (outline)        152014
Triangles (outline)      76371
Triangles (filled)       258682
Rounded rects (outline)  58499
Rounded rects (filled)   730607
Done!
```

**Reply by Westcott1** [reply]
I tried the Adafruit test, but all of the tft.readcommand8 diagnostics return 0x0
I also note that, using a multimeter, rst is always high, and cs is high, but goes low briefly (between graphics draws?).
Which SPI_MODE should be used?

**Reply by Bodmer** [reply]
Mode 0 is correct. The register read looks good, see commemt above..

**Reply by Westcott1** [reply]
Sorry to resurrect this old post, but I've now tried a generic ILI9341 with touch display, and it worked first time.
Waveshare have published a sketch for Arduino Uno, I'll give it a try when I get a Uno.
